notebooks/neural_network_and_data_loading.ipynb

```diff
@@ -0,0 +1,460 @@
+        "        timeout=timeout,\n",
+        "        worker_init_fn=worker_init_fn)\n",
+        "\n",
+        "class FlattenAndCast(object):\n",
+        "  \"\"\"Flattens an image\"\"\"\n",
+        "\n",
+        "  def __call__(self, pic):\n",
+        "    return onp.ravel(onp.array(pic, dtype=np.float32))\n",
+        "\n",
+        "  def __repr__(self):\n",
+        "      return self.__class__.__name__ + '()'"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "l314jsfP4TN4",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "# Define our dataset, using torch datasets\n",
+        "mnist_dataset = MNIST('/tmp/mnist/', download=True, transform=FlattenAndCast())\n",
+        "training_generator = NumpyLoader(mnist_dataset, batch_size=32, num_workers=0)"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "FTNo4beUvb6t",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "# Get the full train dataset (for checking accuracy while training)\n",
+        "train_images = onp.array(mnist_dataset.train_data).reshape(len(mnist_dataset.train_data), -1)\n",
+        "train_labels = one_hot(onp.array(mnist_dataset.train_labels), n_targets)\n",
+        "\n",
+        "# Get full test dataset\n",
+        "mnist_dataset_test = MNIST('/tmp/mnist/', download=True, train=False)\n",
+        "test_images = np.array(mnist_dataset_test.test_data.numpy().reshape(len(mnist_dataset_test.test_data), -1), dtype=np.float32)\n",
+        "test_labels = one_hot(onp.array(mnist_dataset_test.test_labels), n_targets)"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "xxPd6Qw3Z98v",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "### Training Loop"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "X2DnZo3iYj18",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "import time\n",
+        "from tqdm import tqdm\n",
+        "for epoch in range(num_epochs):\n",
+        "  start_time = time.time()\n",
+        "  for x, y in training_generator:\n",
+        "    y = one_hot(y, n_targets)\n",
+        "    params = update(params, x, y)\n",
+        "  epoch_time = time.time() - start_time\n",
+        "\n",
+        "  train_acc = accuracy(params, train_images, train_labels)\n",
+        "  test_acc = accuracy(params, test_images, test_labels)\n",
+        "  print(\"Epoch {} in {:0.2f} sec\".format(epoch, epoch_time))\n",
+        "  print(\"Training set accuracy {}\".format(train_acc))\n",
+        "  print(\"Test set accuracy {}\".format(test_acc))"
+      ],
+      "execution_count": 0,
+      "outputs": []
```

Comment:
The model didn't converge for me, I got:
Epoch 0 in 14.25 sec
Training set accuracy 0.09863333404064178
Test set accuracy 0.0958000048995018
Epoch 1 in 13.68 sec
Training set accuracy 0.11600000411272049
Test set accuracy 0.11710000783205032
Epoch 2 in 12.40 sec
Training set accuracy 0.10001666843891144
Test set accuracy 0.0998000055551529
Epoch 3 in 11.75 sec
Training set accuracy 0.11276666820049286
Test set accuracy 0.11420000344514847
Epoch 4 in 12.25 sec
Training set accuracy 0.09751667082309723
Test set accuracy 0.09740000218153
Epoch 5 in 12.13 sec
Training set accuracy 0.09915000200271606
Test set accuracy 0.10100000351667404
Epoch 6 in 11.80 sec
Training set accuracy 0.09878333657979965
Test set accuracy 0.09550000727176666
Epoch 7 in 11.80 sec
Training set accuracy 0.10064999759197235
Test set accuracy 0.10060000419616699
Epoch 8 in 12.36 sec
Training set accuracy 0.09035000205039978
Test set accuracy 0.08940000087022781
Epoch 9 in 12.74 sec
Training set accuracy 0.10218333452939987
Test set accuracy 0.10110000520944595

which looks like non-convergence. Is there a bug?

Reply:
Sorry, I guess I introduced this bug somewhere along the line. Revised the architecture and now doing better than before the bug.